libraries/abi_generator/include/eos/abi_generator/abi_generator.hpp

```diff
@@ -0,0 +1,633 @@
+#pragma once
+
+#include <set>
+#include <regex>
+#include <algorithm>
+
+#include "clang/Driver/Options.h"
+#include "clang/AST/AST.h"
+#include "clang/AST/ASTContext.h"
+#include "clang/AST/ASTConsumer.h"
+
+#include "clang/Frontend/FrontendPluginRegistry.h"
+#include "clang/AST/AST.h"
+#include "clang/AST/ASTConsumer.h"
+#include "clang/AST/RecursiveASTVisitor.h"
+#include "clang/Frontend/CompilerInstance.h"
+#include "clang/Sema/Sema.h"
+#include "clang/Lex/Preprocessor.h"
+#include "clang/Tooling/Tooling.h"
+#include "clang/Tooling/CommonOptionsParser.h"
+#include "llvm/Support/raw_ostream.h"
+#include <boost/algorithm/string.hpp>
+
+#include <eos/types/AbiSerializer.hpp>
+#include <eos/types/types.hpp>
+#include <fc/io/json.hpp>
+
+using namespace clang;
+using namespace std;
+using namespace clang::tooling;
+namespace cl = llvm::cl;
+
+namespace eos { namespace abi_generator {
+
+FC_DECLARE_EXCEPTION( abi_generation_exception, 999999, "Unable to generate abi" );
+
+#define ABI_ASSERT( TEST, ... ) \
+  FC_EXPAND_MACRO( \
+    FC_MULTILINE_MACRO_BEGIN \
+      if( UNLIKELY(!(TEST)) ) \
+      {                                                                      \
+        if( fc::enable_record_assert_trip )                                  \
+           fc::record_assert_trip( __FILE__, __LINE__, #TEST );              \
+        FC_THROW_EXCEPTION( eos::abi_generator::abi_generation_exception, #TEST ": "  __VA_ARGS__ ); \
+      }                                                                      \
+    FC_MULTILINE_MACRO_END \
+  )
+
+struct AbiGenerator {
+
+  enum optimization {
+    OPT_SINGLE_FIELD_STRUCT
+  };
+
+  bool                   verbose;
+  int                    optimizations;
+  types::Abi*            output;
+  clang::ASTContext*     context;
+  CompilerInstance*      compiler_instance;
+  //map<string, uint64_t>  type_size;
+  map<string, string>    full_types;
+  string                 abi_context;
+
+  AbiGenerator() : optimizations(0), output(nullptr), context(nullptr), compiler_instance(nullptr) {
```

Comment:
FYI. New coding standard is:
* All structs, classes, methods, types should be lower case with _ between words including constants.
* All Template arguments should be CamelCase
* All MACROS should be all UPPERCASE

However, this was all created before that was made clear.

We do need doxygen for all classes and public methods.